blog/how-i-start-rust-2020-03-15.markdown

```diff
@@ -0,0 +1,554 @@
+and see if it works:
+
+```console
+$ ./target/debug/helloworld
+```
+
+And in another terminal window:
+
+```console
+$ curl http://127.0.0.1:8000
+Hello, world!
+$ fg
+<press control-c>
+```
+
+The HTTP service works. We have a binary that is created with the Rust compiler.
+However, it could use some tests.
+
+## Testing
+
+Rocket has support for [unit testing][rockettest] built in. Let's create a tests
+module and verify this route in testing.
+
+[rockettest]: https://rocket.rs/v0.4/guide/testing/
+
+### Create a tests module
+
+Rust allows you to create other [modules][rustmod] inside crates. Create a
+`tests` module that will only build when testing is requested:
+
+[rustmod]: https://doc.rust-lang.org/rust-by-example/mod/visibility.html
+
+```rust
+#[cfg(test)] // Only compile this when unit testing is requested
+mod tests {
+  use super::*; // Modules are their own scope, so you 
+                // need to explictly use the stuff in
+                // the parent module.
+                
+  use rocket::http::Status;
+  use rocket::local::*;
+  
+  #[test]
+  fn test_index() {
+    // create the rocket instance to test
+    let rkt = rocket::ignite().mount("/", routes![index]);
+    
+    // create a HTTP client bound to this rocket instance
+    let client = Client::new(rkt).expect("valid rocket");
+    
+    // get a HTTP response
+    let mut response = client.get("/").dispatch();
+    
+    // Ensure it returns HTTP 200
+    assert_eq!(response.status(), Status::Ok);
+    
+    // Ensure the body is what we expect it to be
+    assert_eq!(response.body_string(), Some("Hello, world!".into()));
+  }
+}
+```
+
+### Run tests
+
+`cargo test` is used to run tests in Rust. Let's run it:
+
+```console
+$ cargo test
+   Compiling helloworld v0.1.0 (/home/cadey/code/helloworld)
+    Finished test [unoptimized + debuginfo] target(s) in 1.80s
+     Running target/debug/deps/helloworld-49d1bd4d4f816617
+
+running 1 test
+test tests::test_index ... ok
+```
+
+## Adding functionality
+
+Most HTTP services return [JSON][json] or JavaScript Object Notation as a way to
+pass objects between computer programs. Let's use Rocket's [JSON
+support][rocketjson] to add a `/hostinfo` route to this app that returns some
+simple information:
+
+[json]: https://www.json.org/json-en.html
+[rocketjson]: https://api.rocket.rs/v0.4/rocket_contrib/json/index.html
+
+- the hostname of the computer serving the response
+- the process ID of the HTTP service
+- the uptime of the system in seconds
+
+### Encoding things to JSON
+
+For encoding things to JSON, we will be using [serde][serde]. We will need to
+add serde as a dependency. Open `Cargo.toml` and put the following lines in it:
+
+[serde]: https://serde.rs/
+
+```toml
+[dependencies]
+serde_json = "1.0"
+serde = { version = "1.0", features = ["derive"] }
+```
+
+This enables serde's derivation support, which will allow us to automate away
```

Comment:
"This lets us use `#[derive(Serialize, Deserialize)]` on our Rust structs, if we import those traits."